### PR TITLE
Improve typing for loops

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #7037

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @justusschock @ananthsub @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -178,24 +178,25 @@ def skip(self) -> bool:
         # until `on_run_start`, we use `limit_train_batches` instead
         return self.done or self.trainer.limit_train_batches == 0
 
-    def connect(self, epoch_loop: TrainingEpochLoop):
+    def connect(self, epoch_loop: Optional[TrainingEpochLoop] = None) -> None:  # type: ignore[override]
```
Comment (awaelchli) [inline pytorch_lightning/loops/fit_loop.py:181]: disagree. Calling connect() without arguments makes no sense. the epoch_loop is required.

File: pytorch_lightning/loops/fit_loop.py [hunk 2]
```diff
@@ -215,13 +216,11 @@ def on_advance_start(self) -> None:
         self.trainer.accumulation_scheduler.on_train_epoch_start(self.trainer, self.trainer.lightning_module)
 
         # stores accumulated grad fractions per batch
-        self.epoch_loop.batch_loop.accumulated_loss = TensorRunningAccum(
-            window_length=self.trainer.accumulate_grad_batches
-        )
+        self.epoch_loop.batch_loop.accumulated_loss.window_length = self.trainer.accumulate_grad_batches
```
Comment (awaelchli) [inline pytorch_lightning/loops/fit_loop.py:219]: This should reset the `TensorRunningAccum`. Either keep as before or add the reset() call.

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 3]
```diff
@@ -98,13 +98,10 @@ def on_run_end(self) -> _OUTPUTS_TYPE:
...
         # this is not necessary as the manual loop runs for only 1 iteration, but just in case
         self.manual_loop._hiddens = None
-        output, self._outputs = self._outputs, None  # free memory
+        output, self._outputs = self._outputs, []  # free memory
+        self._remaining_splits = []
         return output
 
-    def teardown(self) -> None:
-        # release memory
-        self._remaining_splits = None
```
Comment (awaelchli) [inline pytorch_lightning/loops/batch/training_batch_loop.py:106]: we still need to free memory when on_run_end didn't run. That's why I put it here! 
what we forgot here is to call `optimizer_loop.teardown()`

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: `_remaining_splits` is filled in `on_run_start`, which means we are free to clean it in `on_run_end`. why wait until `teardown` to do it?

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
